cmd/edgectl/aes_install.go

```diff
@@ -437,166 +428,97 @@ func (i *Installer) Perform(kcontext string) error {
 
-		i.ShowWrapped("Downloading images. (This may take a minute.)")
+	if err := chartDown.Download(); err != nil {
+		i.ShowWrapped(fmt.Sprintf("-> Failed to download Ambassador Edge Stack"), tryAgain, seeDocs)
+		i.Report("fail_download", ScoutMeta{"err", err.Error()})
+		return errors.Errorf("could not download")
+	}
 
-		if err := i.ShowKubectl("install CRDs", crdManifests, "apply", "-f", "-"); err != nil {
-			i.Report("fail_install_crds")
-			return err
-		}
+	// the AES version we have downloaded
+	downVersion := strings.Trim(chartDown.GetChart().AppVersion, "\n")
 
-		if err := i.ShowKubectl("wait for CRDs", "", "wait", "--for", "condition=established", "--timeout=90s", "crd", "-lproduct=aes"); err != nil {
-			i.Report("fail_wait_crds")
-			return err
-		}
+	chartRel, err := chartDown.GetReleaseMgr(defInstallNamespace, chartValues)
+	defer func() { _ = chartDown.Cleanup() }()
+	if err != nil {
+		i.log.Printf("when obtaining the chart manager: %s", err)
+		i.Report("fail_download", ScoutMeta{"err", err.Error()})
+		return err
+	}
 
-		if err := i.ShowKubectl("install AES", aesManifests, "apply", "-f", "-"); err != nil {
-			i.Report("fail_install_aes")
-			return err
-		}
+	if err := chartRel.Sync(i.ctx); err != nil {
+		i.log.Printf("Failed to sync release: %s", err)
+		i.Report("fail_download", ScoutMeta{"err", err.Error()})
+		return err
+	}
+
+	if !chartRel.IsInstalled() {
+		// Ambassador is not installed at all: perform the installation
+		i.SetMetadatum("AES version being installed", "aes_version", downVersion)
+		i.version = downVersion
+		i.ShowWrapped(fmt.Sprintf("-> Installing the Ambassador Edge Stack %s.", i.version))
 
-		if err := i.ShowKubectl("wait for AES", "", "-n", "ambassador", "wait", "--for", "condition=available", "--timeout=90s", "deploy", "-lproduct=aes"); err != nil {
-			i.Report("fail_wait_aes")
+		installedRelease, err := chartRel.InstallRelease(i.ctx)
+		if err != nil {
+			i.log.Printf("Installation of a new release failed: %s", err)
+			i.Report("fail_install_aes", ScoutMeta{"err", err.Error()})
 			return err
 		}
-	}
 
-	// Wait for Ambassador Pod; grab AES install ID
-	if err := i.loopUntil("AES pod startup", i.GrabAESInstallID, lc2); err != nil {
-		i.Report("fail_pod_timeout")
-		return err
-	}
+		i.log.Printf("New release installed successfully")
+		i.log.Printf("Config values: %+v", installedRelease.Config)
 
-	// Grab Helm information if present
-	if managedDeployment, err := i.CaptureKubectl("get deployment labels", "", "get", "-n", "ambassador", "deployments", "ambassador", "-Lapp.kubernetes.io/managed-by"); err == nil {
-		if strings.Contains(managedDeployment, "Helm") {
-			i.SetMetadatum("Cluster Info", "managed", "helm")
+		message := ""
+		if installedRelease.Info != nil {
+			message = installedRelease.Info.Notes
 		}
+		i.ShowWrapped(message, "\n")
+		i.Report("deploy")
+	} else {
+		// Ambassador is currently installed. Don't do anything.
```

Comment:
What happens if the wrong version of Ambassador is installed? What if there is an incomplete installation? The old code handled those cases (poorly!), but this _appears to_ ignore them completely. I admit I don't know what the Helm library is doing under the covers, so maybe this is okay. Can you clarify?

Reply:
What do you mean with a _wrong version of Ambassador_?

Regarding _incomplete installations_, this uninstalls everything if something goes wrong.